### PR TITLE
[load] dashboard reports | debug mode | copy artifacts sas url

### PR DESCRIPTION
**Pull request by mbhardwaj-msft**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [x] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/load/azext_load/data_plane/load_test_run/custom.py [hunk 1]
```diff
@@ -132,6 +133,86 @@ def stop_test_run(cmd, load_test_resource, test_run_id, resource_group_name=None
     return response
 
 
+def copy_test_run_artifacts_url(cmd, load_test_resource, test_run_id, resource_group_name=None):
+    client = get_testrun_data_plane_client(cmd, load_test_resource, resource_group_name)
+    logger.info("Fetching test run copy artifacts SAS URL for test run %s", test_run_id)
+    test_run_data = client.get_test_run(test_run_id=test_run_id)
+    artifacts_container = test_run_data.get(
+        "testArtifacts", {}).get(
```
Comment (sulabh-msft) [inline src/load/azext_load/data_plane/load_test_run/custom.py:141]: Can we please check if this is safe for negative scenarios e.g. 
- no output artifacts, ( test run in process)
- Test run failed 

- - invalid test run id

File: src/load/azext_load/data_plane/load_test_run/custom.py [hunk 2]
```diff
@@ -132,6 +133,86 @@ def stop_test_run(cmd, load_test_resource, test_run_id, resource_group_name=None
...
+            logger.info("No results file found for test run %s", test_run_id)
+    else:
+        logger.warning(
+            "No results file found for test run %s",
+            test_run_id,
+        )
+
+
+def _download_reports_file(test_run_output_artifacts, test_run_id, path):
+    logger.info("Downloading report file for test run %s", test_run_id)
```
Comment (sulabh-msft) [inline src/load/azext_load/data_plane/load_test_run/custom.py:168]: nit: these methods are doing similar action can we have one method `_download_artifact_file` and remove the duplicate code

Comment (mbhardwaj-msft) [reply src/load/azext_load/data_plane/load_test_run/custom.py:168]: I have refactored as much as possible to reduce duplicate code, the objects that need to be read from parent artifacts object is different in each, the logs which need to be given out are different.

File: src/load/azext_load/data_plane/load_test_run/custom.py (hunk 1 above)
Comment (mbhardwaj-msft) [reply src/load/azext_load/data_plane/load_test_run/custom.py:141]: The SAS URL will be given out even when test run is in progress or test run failed, as long as there are input artifacts associated

File: src/load/azext_load/tests/latest/resources/config.yaml [hunk 3]
```diff
@@ -9,6 +9,9 @@ zipArtifacts:
 failureCriteria:
   - avg(requests_per_sec) > 78
   - percentage(error) > 50
+  - p75(response_time_ms) > 380
+  - p99(response_time_ms) > 520
+  - p99.9(response_time_ms) > 540
```
Comment (sulabh-msft) [inline src/load/azext_load/tests/latest/resources/config.yaml:14]: add percentile for p99.99 etc.

File: src/load/azext_load/tests/latest/test_load_test.py [hunk 4]
```diff
@@ -148,17 +148,29 @@ def test_load_test_create(self, rg, load, vnet):
...
+        # Use a tolerance value to check if the values are close enough.
         for item in pass_fail_metric.values():
             if item.get("clientMetric") == "requests_per_sec":
-                assert item.get("value") == 78.0
+                assert abs(item.get("value") - 78.0) < 1e-6
                 assert item.get("condition") == ">"
                 assert item.get("aggregate") == "avg"
             elif item.get("clientMetric") == "error":
-                assert item.get("value") == 50.0
+                assert abs(item.get("value") - 50.0) < 1e-6
```
Comment (sulabh-msft) [inline src/load/azext_load/tests/latest/test_load_test.py]: move to a const and use it in all these tests ?

File: src/load/azext_load/tests/latest/test_load_testrun.py [hunk 5]
```diff
@@ -351,6 +351,20 @@ def test_load_test_run_download_files(self, rg, load):
...
+            response = self.cmd(
+                "az load test-run show "
+                "--load-test-resource {load_test_resource} "
+                "--resource-group {resource_group} "
+                "--test-run-id {test_run_id}",
+            ).get_output_in_json()
+            if response.get("testArtifacts", {}).get("outputArtifacts", {}).get("reportFileInfo") is not None:
+                break
+            sleep_counter += 1
+            time.sleep(10)
```
Comment (sulabh-msft) [inline src/load/azext_load/tests/latest/test_load_testrun.py:366]: Is this timer sufficient for the test?

**Review by sulabh-msft — APPROVED** [review]

File: src/load/azext_load/data_plane/load_test_run/custom.py (hunk 1 above)
Comment (sulabh-msft) [reply src/load/azext_load/data_plane/load_test_run/custom.py:141]: Is this the case with output artifacts as well?

consider provisioning failed for the VNET scnearios. Will we have the output articats

File: src/load/azext_load/tests/latest/resources/config.yaml (hunk 3 above)
Comment (mbhardwaj-msft) [reply src/load/azext_load/tests/latest/resources/config.yaml:14]: we don't need to add test cases for all, just needed to verify that adding new percentiles in service will work for cli as well directly

File: src/load/azext_load/tests/latest/test_load_testrun.py (hunk 5 above)
Comment (mbhardwaj-msft) [reply src/load/azext_load/tests/latest/test_load_testrun.py:366]: the counter is mostly helpful when we run the tests in live mode, during PR builds and CI/CD, recordings are anyways used and we don't actually depend on the test run actually executing, so this much I believe is good enough

File: src/load/azext_load/tests/latest/test_load_test.py (hunk 4 above)
Comment (mbhardwaj-msft) [reply src/load/azext_load/tests/latest/test_load_test.py]: updated

File: src/load/azext_load/data_plane/load_test_run/custom.py (hunk 1 above)
Comment (mbhardwaj-msft) [reply src/load/azext_load/data_plane/load_test_run/custom.py:141]: added

**Pull request merged**